### PR TITLE
volumes: implement plan diff for volume requests

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/9962

The details of host volume and CSI volume requests do not show up in `nomad
plan` outputs, although the updates are detected by the scheduler and result
in an update as expected.

### CODE REVIEW

**Review by drewbailey — APPROVED** [review]
lgtm!

File: nomad/structs/diff.go [hunk 1]
```diff
@@ -1559,6 +1564,103 @@ Loop:
 	return diff
 }
 
+// volumeDiffs returns the diff of a group's volume requests. If contextual
+// diff is enabled, all fields will be returned, even if no diff occurred.
+func volumeDiffs(old, new map[string]*VolumeRequest, contextual bool) []*ObjectDiff {
```
Comment (shoenig) [inline nomad/structs/diff.go:1569]: Kinda wish we wouldn't shadow `new`

**Review by shoenig — APPROVED** [review]

File: nomad/structs/diff.go (hunk 1 above)
Comment (tgross) [reply nomad/structs/diff.go]: Oof... good call. I've fixed it in this PR but in the interest of not making a huge diff of unrelated style changes I'm not going to fix it where we're doing it everywhere else in `diff.go` in this PR

**Comment by tgross** [comment]
Added a changelog entry (and fixed the ordering). Will merge once we're green

**Pull request merged**